File: DS Problems/Stack/reverseStack.cpp
#include<bits/stdc++.h>
using namespace std;

void display(stack<int> s){
    while(!s.empty()){
        cout<<s.top()<<endl;
        s.pop();
    }
}

void insert(stack<int> &s, int el){
    if(s.empty()){
        s.push(el);
        return;
    }
    int top = s.top();
    s.pop();
    insert(s, el);
    s.push(top);
}

void reverse(stack<int> &s){
    if(s.empty()) return;
    int el = s.top();
    s.pop();
    reverse(s);
    insert(s, el);
}

int main(){
    stack<int> s;
    int i = 1;
    while(i<10) {
        int v = rand()%100;
        cout<<"Pushing: "<<v<<endl;
        s.push(v);
        i++;
    }
    reverse(s);
    display(s);
}

File: DS_Problems/Stack/ReverseStack.java
package DS_Problems.Stack;

import java.util.Stack;

public class ReverseStack {
    public static void main(String[] args) {
        solve();
    }

    public static void solve() {
        Stack<Integer> s = new Stack<>();
        for (int i = 0; i < 5; i++)
            s.add(i);

        reverse(s);

        while (!s.empty()) {
            System.out.println(s.pop());
        }
    }

    static void reverse(Stack<Integer> s) {
        if (s.empty()) {
            return;
        }
        int top = s.pop();
        reverse(s);
        insert(s, top);
    }

    static void insert(Stack<Integer> s, int el) {
        if (s.empty()) {
            s.add(el);
            return;
        }

        int top = s.pop();
        insert(s, el);
        s.add(top);
    }
}
